server/bin/test/test_pbench_server_put_shim_001.py

```diff
@@ -2,135 +2,152 @@
-        crt.copy_result_tb("token")
-    assert str(e).endswith(expected_error_message)
-    """
-    tarball_md5 = "tbmd554"
-    tarball_len = 25
-    controller = "ctrl"
-    tarball = Path(
-        "/home/rriya/Git/pbench/server/bin/test/fixtures/upload/tarball/tar/log.tar.xz"
-    )
-    upload_url = "http://pbench.example.com/api/v1/upload"
-    print("token ======== ", token)
     headers = {
-        "Content-MD5": tarball_md5,
+        "Content-MD5": self.tarball_md5,
         "Authorization": f"Bearer {token}",
-        "controller": controller,
+        "controller": self.controller,
     }
-    with tarball.open("rb") as f:
+    with self.tarball.open("rb") as f:
         try:
             request = requests.Request(
-                "PUT", upload_url, data=f, headers=headers
+                "PUT", self.upload_url, data=f, headers=headers
             ).prepare()
-
-            print("tarball ========== ", tarball)
-            print(
-                "request ========== ",
-                request,
-                "header ====== ",
-                request.headers["Content-Length"],
+            self.logger.info(
+                "File uploaded successfully: {} on target address, {}",
+                self.tarball,
+                self.upload_url,
             )
-            print("tarball_len ========== ", str(tarball_len))
-
-            assert (
-                "Transfer-Encoding" not in request.headers
-            ), "Upload request unexpectedly contains a `Transfer-Encoding` header"
-            assert (
-                "Content-Length" in request.headers
-            ), "Upload request unexpectedly missing a `Content-Length` header"
-
-            print("no ideaaaaaaaaaa")
-            response = requests.Session().send(request)
-            print("response ============ ", response)
-            response.raise_for_status()
-        except requests.exceptions.ConnectionError:
-            raise RuntimeError(f"Cannot connect to '{upload_url}'")
-        except Exception as exc:
-            raise FileUploadError(
-                "There was something wrong with file upload request: "
-                f"file: '{tarball}', URL: '{upload_url}';"
-                f" error: '{exc}'"
-            )
-    assert (
-        response.ok
-    ), f"Logic bomb!  Unexpected error response, '{response.reason}' ({response.status_code})"
+        except Exception:
+            raise RuntimeError(f"Cannot connect to '{self.upload_url}'")
 
 
 class TestProcessTb:
     @pytest.fixture(autouse=True)
-    def config_and_logger(self, server_config):
+    def config_and_logger(self, valid_config):
         # Setup the configuration and logger
-        self.config = server_config
-        self.logger = get_pbench_logger("pbench", server_config)
+        self.config = valid_config
+        self.logger = get_pbench_logger("pbench", valid_config)
+        self.log_file = self.config.LOGSDIR / "pbench" / "pbench.log"
```

Comment:
yeah sure I will try that :D

Reply:
Hey, @webb can you elaborate what you mean by in-memory logger

Reply:
In short, it would be a logger which implements the existing interface but which "writes" messages to a buffer in memory instead of writing them to a file on disk.  I'm not sure exactly what our options are for that.  The logging facility is pretty flexible, as is Python in general (i.e., most places where you need a file, you can instead supply a "file-like object", which can then do what you want).  So, one of us would need to do some research.  But, I think it would be worthwhile, especially for the unit tests, where we either ignore the log or read it immediately -- having the messages go out to disk is extra overhead, and it requires us to set up directories and to clean them up afterwards, none of which is necessary for the task of testing.

Reply:
And why aren't we using `caplog` to layer on top of the `Logger` system instead of guessing where `get_pbench_logger` puts a `file` log in order to read it directly? (E.g, `get_last_log_line`?)

Reply:
My suggestion would be to worry about a different implementation of the logger for unit testing as a separate issue: it should not be part of this PR. OTOH, using caplog  sounds like a great idea.

Reply:
Opened issue #2384 to track this.

Reply:
Thank you for opening another issue to track this